### DISCUSSION TITLE
How to prevent hacking

### DISCUSSION THREAD

**Discussion by thdoan** [Glench/ExtPay#25] [General]
Hi, I am exploring different payment platforms for extensions that are hack-proof (or at least harder to crack). I haven't used ExtPay, yet, but I'm curious to know how you prevent someone from opening `background.js` and changing...

```
extpay.getUser().then(user => {
    if (user.paid) {
        // ...
    } else {
        // ...
    }
})
```

to...

```
extpay.getUser().then(user => {
    // Use paid feature without paying
})
```

Does anyone have a good work-around for this? The only promising solution that I can think of is to use Google's new [Identity Services library](https://developers.google.com/identity/gsi/web/guides/migration), which decouples user authentication and authorization. However, I don't have any experience using the Identity Services library -- is it possible to maintain a list of Google accounts (email addresses) in GCP that are authorized to use an API used by the extension? If this is possible, then even if the user authenticates, since their account is not on the authorized list they won't be able to make the actual API calls used by the extension. The only downside is it only applies to extensions that use GAPI.

**Comment by gannonbarnett** [comment]
Yeah, I'm looking into using ExtPay now and noticed the same thing. The payment UI + auth is useful and I'd like to give it a try. I think to circumvent the above threat, I'll need to have the paid feature created server-side, and also support OAauth between my extension and my server. 

@Glench, or maintainers of ExtPay, could you comment on this vuln?

**Comment by idolize** [comment]
Any client side code in a browser extension can be extracted and run by a sufficiently motivated/technical user regardless of what protections you put in place (short of crazy encryption and/or obfuscation systems like those in modern video games, or measures such as only downloading the JS code from a server after the user has been verified as paid- which is not allowed on the Chrome extension store).

**Reply by Glench** [reply]
This is the right response. Even [Mozilla's article about monetizing extensions](https://extensionworkshop.com/documentation/publish/make-money-from-browser-extensions/) has this to say:

> Developer Tip: Do not spend too much time securing your licensing system against hackers, because users who are inclined to use a hacked license are unlikely to paid for one. Your time is better spent developing new extension features that attract more paying users.

That said, ExtensionPay has taken all the security measures necessary to prevent misuse. This includes not storing user's payment information in the extension's storage in order to prevent easy tampering and limiting the number of simultaneous logins a user gets.

In our experience, most users don't have the technical know-how or motivation to circumvent the system. Statistically-speaking, it is much more likely that no users pay for your extension than someone takes the time to read the source code and hack it, so it's not a huge concern. The fact that ExtensionPay developers have made over $300k as of this writing is testament to that fact.

**Comment by Glench** [comment]
[Mozilla's article about monetizing extensions](https://extensionworkshop.com/documentation/publish/make-money-from-browser-extensions/) has this to say:

> Developer Tip: Do not spend too much time securing your licensing system against hackers, because users who are inclined to use a hacked license are unlikely to paid for one. Your time is better spent developing new extension features that attract more paying users.

That said, ExtensionPay has taken all the security measures necessary to prevent misuse. This includes not storing user's payment information in the extension's storage in order to prevent easy tampering and limiting the number of simultaneous logins a user gets.

In our experience, most users don't have the technical know-how or motivation to circumvent the system. Statistically-speaking, it is much more likely that no users pay for your extension than someone takes the time to read the source code and hack it, so it's not a huge concern. The fact that ExtensionPay developers have made over $300k as of this writing is testament to that fact.
